### ISSUE TITLE
Issue on MSM site: selection being removed for File Field across sites

### ISSUE THREAD

**Issue by sburris777** [ExpressionEngine/ExpressionEngine#78]
When I assign an Allowed Directory for a File Field on one site, it removes that selection on another site. The directories are site specific, but the sites share Channels/Fields. 

![screenshot-lumsdencpa wwwsrc10 supercp com-2019 01 03-16-25-32](https://user-images.githubusercontent.com/24320851/50705659-80ac7480-1029-11e9-90d0-a06bff43f46b.png)

**Comment by robinsowell** [comment]
Related- if a file field is set to allow uploading/access to 'All' directories, on the publish page in the file browser, it only shows directories from that specific site.

If the file field allows uploading/access to a specific directory, on the publish page in the file browser, the browser will be blank and unusable on any MSM site other than the one associated with that specific site.

**Comment by sburris777** [comment]
@robinsowell Is there any update on a fix for this issue or do I need to put in another ticket? I upgraded to 5.1.2, but the issue is still present.

**Comment by robinsowell** [comment]
No, don't duplicate, it will just confuse things.  When there's a pull request or a commit that closes the issue, anyone subscribed to the issue will be notified. So it's in the right spot and once it's resolved, you should get the notification.

**Comment by bigrepository** [comment]
In essence the problem is: Upload directories do not have a "use for all sites" option. So the file field can be shared amongst different sites, but an upload directory can not, since it has the site id saved in the database. 

The file field on the other hand only allows **one** directory to be specified and saves the directory id in it's options, so even defining the same upload destination on the second site, with the same name and the same path will not work, because it's site specific.

We'd need an option when creating an upload destination to set the site id to 0 for "available for all sites". If anyone could look into that, that would be awesome.

**Comment by BridgingUnit** [comment]
I've just run into similar issue which I think is what happens on the other side of this.

I get a very similar error to the one shared by mcfarlandonline over [here](https://github.com/ExpressionEngine/ExpressionEngine/issues/206#issuecomment-492836878)

It seems to be because I'm trying to edit a field within a secondary MSM site that's been shared from the primary site where it was originally set up. It references a File Upload Pref also set up within the Primary site, which alas is not available to the secondary, though you wouldn't know this unless you dug.

I've hit this with both Field and File Grid fields within the same Fluid field.

This issue also has a not necessarily immediately obvious side-effect. You can edit settings for a field within a site (one that's shared with another) for something related to a File Upload Pref, e.g. Allowed Directory on a File field, and you'll be updating settings for the field, and you'll think that's all great and that things are now working fine, but it'll mean that as soon as you hit save on the setting, then all  instances of that field used in other sites won't work, because the setting (as saved in exp_channel_fields within the DB) is then pointing to a file upload preference no longer available to those other sites.

**Comment by adster101** [comment]
I've just run into this bug which IMO is quite nasty. I want to share a file field across sites using a fluid field but currently that's not possible.

Not sure I agree with @bigrepository suggestion about adding 'available for all sites' option to the upload prefs. Also I'm not sure that the workaround suggested by @BridgingUnit actually works in my case. 

Would changing the file fieldtype so it works across multiple sites be a viable option?

Currently the settings are saved to 'exp_channel_fields.field_settings' as an array like below. 

array (size=5)
  'field_content_type' => string 'all' (length=3)
  'allowed_directories' => string '6' (length=1)
  'show_existing' => string 'y' (length=1)
  'num_existing' => string '50' (length=2)
  'field_fmt' => string 'none' (length=4)

If field_settings was saved as a nested array it might be possible to save settings for multiple sites. For example the field_settings array might become:

array (size=5)
  'field_content_type' => string 'all' (length=3)
  'allowed_directories' => array(6,3) (size=2) <-- Upload directory IDs for different sites
  'show_existing' => string 'y' (length=1)
  'num_existing' => string '50' (length=2)
  'field_fmt' => string 'none' (length=4)

Obviously we'd need to update the save_settings function in ft.file.php and various others but could this work? Would it be a sensible approach to correct the behaviour of this field?

**Comment by adster101** [comment]
**Update**: The workaround suggested by @BridgingUnit does work assuming you're on the latest version of EE (5.3.2). Having  just upgraded my site and choosing 'all' for the file field seems to work as expected. With only one upload directory this would seem to work more or less as expected. 

Still happy to look at the above fix if it helps.

**Comment by intoeetive** [comment]
Digging into this, I do like the approach @bigrepository suggests, however the problem is also that uploaded files are associated with site as well, not just upload directory. So if we make the upload destinations shared, it might end up with extra confusion when some files will be visible on one MSM site, but not the other.

Another approach would be keeping upload destinations per-site, but allowing File field to use appropriate destinations when used on different MSM sites. But currently File field does not support multiple destinations (which has been requested in https://github.com/ExpressionEngine/ExpressionEngine/issues/928)

**Comment by robsonsobral** [comment]
And... What if the field uses the directory name, but still saves the ID? I mean, at form loading, it gets the directory of based on the fieldset? So, if the directories share names, the field don't break. Too bad idea?

**Comment by jamesseavers** [comment]
Can this fix be rolled into EE6 also?
